R/discrete_normal.R

```diff
@@ -0,0 +1,118 @@
+#'
+#' @importFrom R6 R6Class
+#' @export
+
+discrete_normal <- function(mean, sd, breaks, dim = NULL) {
+  distrib("discrete_normal", mean, sd, breaks, dim)
+}
+
+# define the discrete normal distribution
+discrete_normal_distribution <- R6Class(
+  classname = "discrete_normal_distribution",
+  inherit = distribution_node,
+  public = list(
+    
+    breaks = NA,
+    lower_bounds = NA,
+    upper_bounds = NA,
+    
+    initialize = function(mean, sd, breaks, dim) {
+      
+      mean <- as.greta_array(mean)
+      sd   <- as.greta_array(sd)
+      
+      # check length of breaks
+      if (length(breaks) <= 1) {
+        msg <- cli::format_error(
+          c(
+            "{.var breaks} must be a vector with at least two break points",
+            "but {.var breaks} has length {length(breaks)}"
+          )
+        ) 
+        stop(
+          msg,
+          call. = FALSE
+        )
+      }
+      
+      # handle gradient issue between sdlog and 0s
+      breaks <- pmax(breaks, .Machine$double.eps)
+      self$breaks <- breaks
+      self$lower_bounds <- breaks[-length(breaks)]
+      self$upper_bounds <- breaks[-1]
+      
+      # add the nodes as parents and parameters
+      dim <- check_dims(mean, sd, target_dim = dim)
+      super$initialize("discrete_normal", dim, discrete = TRUE)
+      self$add_parameter(mean, "mean")
+      self$add_parameter(sd, "sd")
+      
+    },
+    
+    tf_distrib = function(parameters, dag) {
+      
+      mean <- parameters$mean
+      sd <- parameters$sd
+      
+      tf_breaks <- fl(self$breaks)
+      tf_lower_bounds <- fl(self$lower_bounds)
+      tf_upper_bounds <- fl(self$upper_bounds)
+      
+      log_prob <- function(x) {
+        
+        # build distribution object
+        d <- tfp$distributions$Normal(
+          loc = mean,
+          scale = sd
+        )
+        
+        # for those lumped into groups,
+        # compute the bounds of the observed groups
+        # and get tensors for the bounds in the format expected by TFP
+        x_safe <- tf$math$maximum(x, fl(.Machine$double.eps))
+        tf_idx <- tfp$stats$find_bins(x_safe, tf_breaks)
+        tf_idx_int <- greta:::tf_as_integer(tf_idx)
+        tf_lower_vec <- tf$gather(tf_lower_bounds, tf_idx_int)
+        tf_upper_vec <- tf$gather(tf_upper_bounds, tf_idx_int)
+        
+        # compute the density over the observed groups
+        # note-to-self: this looks like https://mc-stan.org/docs/2_29/stan-users-guide/bayesian-measurement-error-model.html#rounding
+        low <- tf_safe_cdf(tf_lower_vec, d)
+        up <- tf_safe_cdf(tf_upper_vec, d)
+        log_density <- log(up - low)
+        
+      }
+      
+      sample <- function(seed) {
+        
+        d <- tfp$distributions$Normal(
+          loc = mean,
+          scale = sd
+        )
+        continuous <- d$sample(seed = seed)
+        # tf$floor(continuous)
+        tf$round(continuous)
```

Comment:
This assumes that the breaks are all the integers. I think it should instead be `tf$gather()`ed at the breaks. I think this is also a mistake in my implementation of the discrete lognormal (though I never used the sampling in that code).

Reply:
For rounded data, we assume that a user will specify breaks like `seq(from, to, by = 1)`. But for general use we also want to allow something irregular like `c(0, 1, 10, 100)` to work. For the latter, I'm not sure how to gather a continuous variable at the breaks. For instance, do we gather 50.67 into the 10 or the 100 break point? Does this entail an additional argument for `discrete_normal` to gather values into the lower or upper break point?

Furthermore, I like the gather idea because it allows the discretisation into non-integer breaks like `c(-0.2, -0.1, 0, 0.1, 0.2)`, whereas `floor` or `round` discretises into interger only (without tweaks as suggested [here](https://github.com/greta-dev/greta/issues/532)).

Reply:
In [49f6d0e](https://github.com/greta-dev/greta.distributions/pull/16/commits/49f6d0ec6d2f1980a0ce8055cbfc92121f129867) I tried my best to ditto from the `tf$gather()` chunk for `breaks` to discretise the continuous samples into user-defined `breaks`, rather than using `round` or `floor`. I thought this might solve the weird +0.5 estimation bias, but it didn't... But at least we know it isn't due to round vs. gather (if I coded it correctly).